File: agentkit/ts/src/tools/planner.ts
/**
 * ToolPlanner: concurrent scatter-gather execution of LLM tool calls.
 * Mirrors `agentkit.runtime.agents.planner` from the Python SDK.
 */
import { AgentKitEvent } from "../events/schemas";
import { EventType } from "../events/types";
import type { ToolSpec } from "./registry";
import type { ToolPolicy } from "./policy";

const JSON_TYPE_CHECK: Record<string, (v: unknown) => boolean> = {
  object: (v) => typeof v === "object" && v !== null && !Array.isArray(v),
  array: Array.isArray,
  string: (v) => typeof v === "string",
  integer: (v) => typeof v === "number" && Number.isInteger(v),
  number: (v) => typeof v === "number",
  boolean: (v) => typeof v === "boolean",
  null: (v) => v === null,
};

/** Shallow JSON Schema check against ToolSpec.parameters. Returns null on success. */
function validateArgs(spec: ToolSpec, args: Record<string, unknown>): string | null {
  const schema = (spec.parameters ?? {}) as Record<string, unknown>;
  if (
    schema.type === "object" &&
    (typeof args !== "object" || args === null || Array.isArray(args))
  ) {
    return `arguments must be an object, got ${Array.isArray(args) ? "array" : typeof args}`;
  }
  const required = (schema.required as string[] | undefined) ?? [];
  for (const key of required) {
    if (!(key in args)) return `missing required argument: '${key}'`;
  }
  const properties = (schema.properties as Record<string, { type?: string }> | undefined) ?? {};
  for (const [key, propSchema] of Object.entries(properties)) {
    if (!(key in args)) continue;
    const expected = propSchema?.type;
    if (!expected) continue;
    const check = JSON_TYPE_CHECK[expected];
    if (check && !check(args[key])) {
      return `argument '${key}': expected ${expected}, got ${typeof args[key]}`;
    }
  }
  return null;
}

const ERROR_MSG_MAX = 200;

/** Return a log-safe error string. Includes class name and a length-capped message. */
function sanitizeError(err: unknown): string {
  if (err instanceof Error) {
    const msg =
      err.message.length > ERROR_MSG_MAX ? err.message.slice(0, ERROR_MSG_MAX) + "…" : err.message;
    return msg ? `${err.name}: ${msg}` : err.name;
  }
  const s = String(err);
  return s.length > ERROR_MSG_MAX ? s.slice(0, ERROR_MSG_MAX) + "…" : s;
}

/** A single tool call instruction from the LLM. */
export interface ToolCallInstruction {
  id?: string;
  name: string;
  arguments: Record<string, unknown>;
}

/** Result of a single tool call execution. */
export type ToolCallResult =
  | { id: string; name: string; result: unknown }
  | { id: string; name: string; error: string };

export type ToolExecutor = (name: string, args: Record<string, unknown>) => Promise<unknown>;
export type ApprovalHandler = (
  name: string,
  args: Record<string, unknown>,
  risk: string | undefined,
) => Promise<boolean>;
export type EmitFn = (event: AgentKitEvent) => Promise<void>;

export interface ToolPlannerOptions {
  executor: ToolExecutor;
  policy?: ToolPolicy;
  approvalHandler?: ApprovalHandler;
  specs?: Map<string, ToolSpec>;
}

export class ToolPlanner {
  private readonly executor: ToolExecutor;
  private readonly policy: ToolPolicy | undefined;
  private readonly approvalHandler: ApprovalHandler | undefined;
  private readonly specs: Map<string, ToolSpec>;

  constructor(opts: ToolPlannerOptions) {
    this.executor = opts.executor;
    this.policy = opts.policy;
    this.approvalHandler = opts.approvalHandler;
    this.specs = opts.specs ?? new Map();
  }

  /**
   * Execute all tool calls concurrently (scatter-gather) and collect results.
   *
   * Each call emits `TOOL_CALL_REQUESTED` as its first step before entering
   * the approval/execution phase. For multiple simultaneous calls, lifecycle
   * events from different calls will interleave — e.g. call A's
   * `TOOL_CALL_STARTED` may appear between call B's `TOOL_CALL_REQUESTED` and
   * `TOOL_CALL_STARTED`. Consumers that require strict per-call ordering
   * should correlate events by `tool_call_id`.
   *
   * Approval rejection emits both `TOOL_APPROVAL_REJECTED` and
   * `TOOL_CALL_FAILED` so `replaySession` projects the outcome into
   * model-visible history, preventing the agent from re-requesting the same
   * tool indefinitely.
   */
  async executeScatterGather(
    sessionId: string,
    toolCalls: ToolCallInstruction[],
    emit: EmitFn,
  ): Promise<ToolCallResult[]> {
    return Promise.all(toolCalls.map((call) => this.executeSingle(sessionId, call, emit)));
  }

  private async executeSingle(
    sessionId: string,
    call: ToolCallInstruction,
    emit: EmitFn,
  ): Promise<ToolCallResult> {
    const toolName = call.name;
    const toolArgs = call.arguments;
    const callId = call.id ?? crypto.randomUUID();
    const spec = this.specs.get(toolName);
    const risk = spec?.risk;
    const catalogName = spec?.catalogName;
    const aliases = catalogName ? [catalogName] : [];
    const metadata = catalogName ? { catalog_name: catalogName } : {};

    // 1. Announce intent
    await emit(
      AgentKitEvent.parse({
        type: EventType.TOOL_CALL_REQUESTED,
        session_id: sessionId,
        tool_name: toolName,
        tool_args: toolArgs,
        tool_call_id: callId,
        metadata,
      }),
    );

    // 2a. Provider parse-error sentinel: model produced unparseable tool JSON.
    if (typeof toolArgs.__parse_error === "string") {
      const errorMsg = `Invalid tool arguments: ${toolArgs.__parse_error}`;
      await emit(
        AgentKitEvent.parse({
          type: EventType.TOOL_CALL_FAILED,
          session_id: sessionId,
          tool_name: toolName,
          tool_call_id: callId,
          error: errorMsg,
          metadata,
        }),
      );
      return { id: callId, name: toolName, error: errorMsg };
    }

    // 2b. Schema validation: fail closed on malformed args from the model.
    if (spec !== undefined) {
      const schemaError = validateArgs(spec, toolArgs);
      if (schemaError !== null) {
        const errorMsg = `Invalid tool arguments: ${schemaError}`;
        await emit(
          AgentKitEvent.parse({
            type: EventType.TOOL_CALL_FAILED,
            session_id: sessionId,
            tool_name: toolName,
            tool_call_id: callId,
            error: errorMsg,
            metadata,
          }),
        );
        return { id: callId, name: toolName, error: errorMsg };
      }
    }

    // 3. Allow/deny gate: policy violations fail before approval/execution.
    if (this.policy !== undefined && !this.policy.isAllowedAny([toolName, ...aliases])) {
      const errorMsg = `Tool not permitted: ${toolName}`;
      await emit(
        AgentKitEvent.parse({
          type: EventType.TOOL_CALL_FAILED,
          session_id: sessionId,
          tool_name: toolName,
          tool_call_id: callId,
          error: errorMsg,
          metadata,
        }),
      );
      return { id: callId, name: toolName, error: errorMsg };
    }

    // 3. Approval gate
    if (this.policy?.requiresApproval(toolName, risk)) {
      await emit(
        AgentKitEvent.parse({
          type: EventType.TOOL_APPROVAL_REQUESTED,
          session_id: sessionId,
          tool_name: toolName,
          tool_call_id: callId,
          tool_args: toolArgs,
          risk: risk ?? null,
          metadata,
        }),
      );

      let approved = false;
      if (this.approvalHandler !== undefined) {
        approved = await this.approvalHandler(toolName, toolArgs, risk);
      }

      if (!approved) {
        const reason =
          this.approvalHandler === undefined
            ? "No approval handler registered"
            : "Rejected by approval handler";
        const errorMsg = `Tool approval rejected: ${reason}`;
        await emit(
          AgentKitEvent.parse({
            type: EventType.TOOL_APPROVAL_REJECTED,
            session_id: sessionId,
            tool_name: toolName,
            tool_call_id: callId,
            reason,
            metadata,
          }),
        );
        // Also emit TOOL_CALL_FAILED so replaySession projects this into
        // model-visible history. Without it, the next iteration sees no tool
        // result and re-requests the same tool until maxToolIterations.
        await emit(
          AgentKitEvent.parse({
            type: EventType.TOOL_CALL_FAILED,
            session_id: sessionId,
            tool_name: toolName,
            tool_call_id: callId,
            error: errorMsg,
            metadata,
          }),
        );
        return { id: callId, name: toolName, error: errorMsg };
      }

      await emit(
        AgentKitEvent.parse({
          type: EventType.TOOL_APPROVAL_APPROVED,
          session_id: sessionId,
          tool_name: toolName,
          tool_call_id: callId,
          metadata,
        }),
      );
    }

    // 4. Mark execution started
    await emit(
      AgentKitEvent.parse({
        type: EventType.TOOL_CALL_STARTED,
        session_id: sessionId,
        tool_name: toolName,
        tool_call_id: callId,
        metadata,
      }),
    );

    // 5. Execute
    try {
      const result = await this.executor(toolName, toolArgs);
      await emit(
        AgentKitEvent.parse({
          type: EventType.TOOL_CALL_COMPLETED,
          session_id: sessionId,
          tool_name: toolName,
          tool_call_id: callId,
          result,
          metadata,
        }),
      );
      return { id: callId, name: toolName, result };
    } catch (err) {
      const errorMsg = sanitizeError(err);
      await emit(
        AgentKitEvent.parse({
          type: EventType.TOOL_CALL_FAILED,
          session_id: sessionId,
          tool_name: toolName,
          tool_call_id: callId,
          error: errorMsg,
          metadata,
        }),
      );
      return { id: callId, name: toolName, error: errorMsg };
    }
  }
}

File: agentkit/sdk/agentkit/runtime/agents/planner.py
import asyncio
import logging
import uuid
from typing import Any, Awaitable, Callable, Dict, List, Optional

from agentkit.infra.events.schemas import (
    ToolApprovalApproved,
    ToolApprovalRejected,
    ToolApprovalRequested,
    ToolCallCompleted,
    ToolCallFailed,
    ToolCallRequested,
    ToolCallStarted,
)
from agentkit.runtime.tools.policies import ToolPolicy, ToolPolicyViolation
from agentkit.runtime.tools.registry import ToolSpec

logger = logging.getLogger(__name__)

ToolExecutor = Callable[[str, Dict[str, Any]], Awaitable[Any]]
# Approval handler: receives (tool_name, tool_args, risk) and returns True to approve.
ApprovalHandler = Callable[[str, Dict[str, Any], Optional[str]], Awaitable[bool]]


class ToolPlanner:
    """Plans and executes tool calls concurrently (Scatter-Gather).

    Args:
        executor: An async callable ``(tool_name: str, args: dict) -> dict``
            that dispatches a single tool call. Typically wraps ``execute_tool``
            from the tool registry:

                ToolPlanner(
                    executor=lambda name, args: execute_tool("user-1", name, args)
                )

            For per-agent scoping with a ``ToolRegistry`` instance:

                ToolPlanner(
                    executor=lambda name, args: registry.execute("user-1", name, args)
                )

        policy: Optional ``ToolPolicy``. When provided, tools whose risk level
            is in ``policy.require_approval_for`` will pause for approval via
            ``approval_handler`` before execution.

        approval_handler: Async callback invoked when a tool requires approval.
            Receives ``(tool_name, tool_args, risk)`` and must return ``True``
            to allow execution or ``False`` to reject it. When omitted and a
            tool requires approval, it is rejected by default (fail-safe).

        specs: Optional mapping of tool name → ``ToolSpec`` used to look up
            the ``risk`` field per tool. When not provided, risk is treated as
            ``None`` (unclassified) for all tools.
    """

    def __init__(
        self,
        executor: ToolExecutor,
        *,
        policy: Optional[ToolPolicy] = None,
        approval_handler: Optional[ApprovalHandler] = None,
        specs: Optional[Dict[str, ToolSpec]] = None,
    ):
        self.executor = executor
        self.policy = policy
        self.approval_handler = approval_handler
        self._specs: Dict[str, ToolSpec] = specs or {}

    async def execute_scatter_gather(
        self,
        session_id: str,
        tool_calls: List[Dict[str, Any]],
        emit_event: Callable[[Any], Awaitable[None]],
    ) -> List[Dict[str, Any]]:
        """Execute multiple tools simultaneously and emit lifecycle events."""
        tasks = []
        for call in tool_calls:
            tasks.append(self._execute_single(session_id, call, emit_event))

        return list(await asyncio.gather(*tasks))

    async def _execute_single(
        self,
        session_id: str,
        call: Dict[str, Any],
        emit_event: Callable[[Any], Awaitable[None]],
    ) -> Dict[str, Any]:
        """Execute a single tool with policy enforcement and approval hooks."""
        tool_name = call.get("name", "unknown")
        tool_args = call.get("arguments", {})
        call_id = call.get("id", str(uuid.uuid4()))
        spec = self._specs.get(tool_name)
        risk = spec.risk if spec else None

        # 1. Announce intent to call
        await emit_event(
            ToolCallRequested(
                session_id=session_id,
                tool_name=tool_name,
                tool_args=tool_args,
                tool_call_id=call_id,
            )
        )

        # 2. Allow/deny gate: policy violations fail before approval/execution.
        if self.policy is not None:
            try:
                self.policy.enforce(tool_name)
            except ToolPolicyViolation as error:
                error_msg = str(error)
                await emit_event(
                    ToolCallFailed(
                        session_id=session_id,
                        tool_name=tool_name,
                        tool_call_id=call_id,
                        error=error_msg,
                    )
                )
                return {"id": call_id, "name": tool_name, "error": error_msg}

        # 3. Approval gate: if policy requires it, pause and ask.
        if self.policy is not None and self.policy.requires_approval(tool_name, risk):
            await emit_event(
                ToolApprovalRequested(
                    session_id=session_id,
                    tool_name=tool_name,
                    tool_call_id=call_id,
                    tool_args=tool_args,
                    risk=risk,
                )
            )

            approved = False
            if self.approval_handler is not None:
                approved = await self.approval_handler(tool_name, tool_args, risk)

            if not approved:
                reason = (
                    "No approval handler registered"
                    if self.approval_handler is None
                    else "Rejected by approval handler"
                )
                await emit_event(
                    ToolApprovalRejected(
                        session_id=session_id,
                        tool_name=tool_name,
                        tool_call_id=call_id,
                        reason=reason,
                    )
                )
                return {
                    "id": call_id,
                    "name": tool_name,
                    "error": f"Tool approval rejected: {reason}",
                }

            await emit_event(
                ToolApprovalApproved(
                    session_id=session_id,
                    tool_name=tool_name,
                    tool_call_id=call_id,
                )
            )

        # 4. Mark execution as started
        await emit_event(
            ToolCallStarted(
                session_id=session_id, tool_name=tool_name, tool_call_id=call_id
            )
        )

        try:
            # 5. Call the actual implementation
            result = await self.executor(tool_name, tool_args)

            # 6. Mark success
            await emit_event(
                ToolCallCompleted(
                    session_id=session_id,
                    tool_name=tool_name,
                    tool_call_id=call_id,
                    result=result,
                )
            )
            return {"id": call_id, "name": tool_name, "result": result}

        except Exception as e:
            error_msg = str(e)
            logger.error("Tool execution failed: %s", error_msg)

            # 6. Mark failure
            await emit_event(
                ToolCallFailed(
                    session_id=session_id,
                    tool_name=tool_name,
                    tool_call_id=call_id,
                    error=error_msg,
                )
            )
            return {"id": call_id, "name": tool_name, "error": error_msg}
